Elasticsearch

liamwhite opened the pull request.

Commit:
add post index

Commit:
first pass, mostly works

Commit:
add elasticsearch to vagrant setup

liamwhite committed:
apparently can't spell system

liamwhite committed:
add supposedly faster indexing method

liamwhite committed:
add indexing jobs

Commit:
Add pagination support to Elasticsearch

Commit:
Fix some references to tag_match and broken index jobs

liamwhite committed:
add random sort to query builder

liamwhite committed:
cleanups

liamwhite committed:
fix post import bugs

Commit:
Better paginator code

This paginator materializes the output into an array and gives you
the option of records or results as needed.

Updated a few references to raw_tag_match and fixed raw_tag_match
code.

Commit:
Add back limit on tag count, raised defaults

In order to prevent unbounded term generation, add back tag count
limit to the query generator.

Updated defaults to reflect the great increase in speed.

liamwhite committed:
mistakenly removed

Commit:
Add pool and set searching

Fix query generator nesting of should items

Commit:
A whole boatload of elasticsearch changes

Makes index job unique to account for changes to update triggering.
Fixes instances of mixed SQL and ES searches with tag_matches.
Adds utility methods for exists? and count_only ES searches.
Adds comment_bumped_at for comments page usage.
Adds random seeding capability.
Fixes random searches not working correctly.
Fixes status:active.

Commit:
Fixes count_only returning wrong result

Commit:
Add back SQL based tag search

In order to avoid making some terrible batch loading of posts by
ids for ES, add back the existing tag search method. This can
be removed at a later time when a batched loader is added.

This also fixes a few instances of count methods.

liamwhite commented on app/logical/post_query_builder.rb:
```diff
@@ -502,10 +507,19 @@ def should(*args)
-          query: {match_all: {}},
-          random_score: {},
-      }})
+      if q[:random].present?
```
I'd recommend
```rb
        must.push({function_score: {
            query: {match_all: {}},
            random_score: {seed: q[:random].to_i},
            boost_mode: :replace
        }})
```
for both cases. (Note that `NilClass#to_i` is defined and always returns zero.)

kiranoot replied:
I need to revise this anyways, because if you specify a seed it uses a different sorting field than if you don't specify one. So splitting them does make sense. For the future I need to specify a field as well and set it to the `_seq_no` field or the `id` field so that it doesn't load larger less efficient fields into memory.

liamwhite commented on app/logical/danbooru/paginator/elasticsearch_extensions.rb:
```diff
@@ -162,6 +162,16 @@ def total_count
       end
+
+      def exists?
+        search.definition.update(from: 0, size: 0, terminate_after: 1)
```
I've found that to actually get the fastest results for an existence query, you actually want these options:
`{ sort: '_doc', _source: false, size: 1, track_total_hits: false, terminate_after: 1 }`

liamwhite commented on app/logical/related_tag_calculator.rb:
```diff
@@ -38,9 +38,9 @@ def self.calculate_similar_from_sample(tag)
     similar_counts = Hash.new {|h, k| h[k] = 0}
     CurrentUser.without_safe_mode do
-      PostReadOnly.with_timeout(5_000, nil, {:tags => tag}) do
+      Post.with_timeout(5_000, nil, {:tags => tag}) do
```
Is there a point to having the timeout, since this now barely stresses the database adapter?

Note also that elasticsearch-rails sets a default timeout of 5 seconds for its HTTP client. You can modify this for e.g. indexing purposes like this:
```rb
Elasticsearch::Model.client = Elasticsearch::Client.new(request_timeout: 30)
```

kiranoot replied:
I haven't had a chance to test these, but I don't see a reason to explicitly remove the timeout code since it should act as an extra safety net. I was also debating if it made sense to add in timeouts to the elasticsearch query, or if that would be wasted effort given that it should be very efficient.

Commit:
Silly mistakes in takedown code

Commit:
Fix moderation queue queries

Fix user deletion query

Allow using includes on elasticsearch records calls.

Commit:
Missing search options and some clean up.

Commit:
Adds voted search, fixes vote searches

Commit:
Fix race condition in favorites and voting on posts

This resolves the case where the model update job would be triggered
too soon and miss the committed state for favorites and votes on posts.

The cause of this was larger transaction blocks surrounding updates
and the jobs being queued on the first column update before commit.

This ensures that the post is fully committed before it is queued
for an index update and that all changes are visible in the table.

Commit:
Fix indexing race conditions in sets and pools

liamwhite commented on app/models/favorite.rb:
```diff
@@ -1,5 +1,6 @@
 class Favorite < ApplicationRecord
-  class Error < Exception ; end
+  class Error < Exception;
```
extra semicolon

liamwhite commented on app/models/favorite.rb:
```diff
@@ -8,15 +9,21 @@ class Error < Exception ; end
     ckey = "fav:#{user.id}:#{post.id}"
     raise Error.new("You are already favoriting") if !Cache.get(ckey).nil?
-    Cache.put(ckey, true, 30)
+    begin
```
What's this for?

liamwhite commented on app/models/favorite.rb:
```diff
@@ -27,9 +27,11 @@ def self.add!(post:, user:)
-      Favorite.create!(:user_id => user.id, :post_id => post.id)
-      Post.where(:id => post.id).update_all("fav_count = fav_count + 1")
-      post.append_user_to_fav_string(user.id)
+      post.with_lock do
```
This entire method is pretty much broken.
- `favorites` should have a unique index on `post_id, user_id` (you can remove the separate `post_id` index when you add this)
- The transaction needs to have `serializable` isolation (`Post.transaction(isolation: :serializable)`)
- No other locking is necessary

liamwhite commented on app/models/favorite.rb:
```diff
@@ -51,18 +53,21 @@ def self.remove!(user:, post: nil, post_id: nil)
-      Favorite.for_user(user.id).where(post_id: post_id).delete_all
-      Post.where(:id => post_id).update_all("fav_count = fav_count - 1")
-      post.delete_user_from_fav_string(user.id) if post
+
```
Ditto here

liamwhite commented on app/models/post_vote.rb:
```diff
@@ -38,18 +38,30 @@ def initialize_attributes
-      Post.where(:id => post_id).update_all("score = score + #{score}, up_score = up_score + #{score}")
-    else
-      Post.where(:id => post_id).update_all("score = score + #{score}, down_score = down_score + #{score}")
+    post = Post.find(post_id)
```
Probably more unnecessary locking here (do an atomic increment and reload)

Commit:
Moved favs/voting out into managers to consolidate logic

While the logic was not particularly complex or spread out it was
made more complicated by the models being dependant on one another.

This creates a central entry point for voting and favorites and
coordinates all transactions.

TODO: upvote/downvote/fav metatags in tag section result in an error
because they attempt to change the transaction isolation mode during
an outer transaction.
